cmd/daemonset-check-refactor/main.go

```diff
@@ -0,0 +1,286 @@
+	ctxCancel context.CancelFunc
+
+	// Cleanup context
+	cleanUpCtx       context.Context
+	cleanUpCtxCancel context.CancelFunc
+
+	// Interrupt signal channels.
+	signalChan chan os.Signal
+	doneChan   chan error
+
+	// K8s client used for the check.
+	client *kubernetes.Clientset
+)
+
+const (
+	// Default k8s manifest resource names.
+	defaultCheckDSName = "daemonset"
+	// Default namespace daemonset check will be performed in
+	defaultCheckNamespace = "kuberhealthy"
+	// Default check pod timeout for daemonset check
+	defaultCheckPodTimeout = time.Duration(time.Minute*12)
+	// Default pause container image used for the daemonset check
+	defaultDSPauseContainerImage = "gcr.io/google-containers/pause:3.1"
+	// Default shutdown termination grace period
+	defaultShutdownGracePeriod = time.Duration(time.Minute * 1) // grace period for the check to shutdown after receiving a shutdown signal
+	// Default daemonset check time limit
+	defaultCheckTimeLimit = time.Duration(time.Minute * 15)
+
+	// Default user
+	defaultUser = int64(1000)
+
+)
+
+func init() {
+
+	// Parse all incoming input environment variables and crash if an error occurs
+	// during parsing process.
+	parseInputValues()
+
+	// Create a timestamp reference for the daemonset;
+	// also to reference against daemonsets that should be cleaned up.
+	now = time.Now()
+	setCheckConfigurations(now)
+
+	// Allocate channels.
+	signalChan = make(chan os.Signal, 5)
+	doneChan = make(chan error, 5)
+}
+
+func main() {
+
+	log.Debugln("Allowing this check", checkTimeLimit, "to finish.")
+	ctx, ctxCancel = context.WithTimeout(context.Background(), checkTimeLimit)
+
+	// Create a kubernetes client.
+	var err error
+	client, err = kubeClient.Create(kubeConfigFile)
+	if err != nil {
+		errorMessage := "failed to create a kubernetes client with error: " + err.Error()
+		reportErrorsToKuberhealthy([]string{"kuberhealthy/daemonset: " + errorMessage})
+		return
+	}
+	log.Infoln("Kubernetes client created.")
+
+
+	// Start listening to interrupts.
+	go listenForInterrupts()
+
+	// Catch panics.
+	var r interface{}
+	defer func() {
+		r = recover()
+		if r != nil {
+			log.Infoln("Recovered panic:", r)
+			reportErrorsToKuberhealthy([]string{"kuberhealthy/daemonset: " + r.(string)})
+		}
+	}()
+
+	//checkPodTimeout = time.Duration(1) * time.Second
+	// Start daemonset check.
+	runCheck()
+}
+
+
+// setCheckConfigurations sets Daemonset configurations
+func setCheckConfigurations(now time.Time) {
+	hostName = getHostname()
+	daemonSetName = checkDSName + "-" + hostName + "-" + strconv.Itoa(int(now.Unix()))
+}
+
+// listenForInterrupts watches the signal and done channels for termination.
+func listenForInterrupts() {
+
+	// Relay incoming OS interrupt signals to the signalChan
+	signal.Notify(signalChan, os.Interrupt, os.Kill)
+	sig :=<-signalChan
+	log.Infoln("Received an interrupt signal from the signal channel.")
+	log.Debugln("Signal received was:", sig.String())
+
+	go shutdown(doneChan)
```

Comment:
i think it would be prudent to pass the context to this function and call the `ctxCancel` func here -- most of your check run steps respect a `ctx.Done()` which would return a result if the context was canceled